services/web/server/src/simcore_service_webserver/meta_db.py

```diff
@@ -0,0 +1,479 @@
+                .fetch("id name head_commit_id")
+            )
+            return branch
+
+    async def _get_head_commit(
+        self, repo_id: int, conn: SAConnection
+    ) -> Optional[RowProxy]:
+        if branch := await self._get_head_branch(repo_id, conn):
+            commit = (
+                await self.CommitsOrm(conn).set_filter(id=branch.head_commit_id).fetch()
+            )
+            return commit
+
+    async def _update_state(self, repo_id: int, conn: SAConnection):
+
+        head_commit: Optional[RowProxy] = await self._get_head_commit(repo_id, conn)
+
+        # current repo
+        repo_orm = self.ReposOrm(conn).set_filter(id=repo_id)
+        returning_cols = "id project_uuid project_checksum modified"
+        repo = await repo_orm.fetch(returning_cols)
+        assert repo  #  nosec
+
+        # fetch project
+        project_orm = self.ProjectsOrm(conn).set_filter(uuid=repo.project_uuid)
+        project = await project_orm.fetch("last_change_date workbench ui")
+        assert project  # nosec
+
+        checksum: Optional[SHA1Str] = repo.project_checksum
+        if not checksum or (checksum and repo.modified < project.last_change_date):
+            checksum = eval_checksum(project.workbench)
+            repo = await repo_orm.update(returning_cols, project_checksum=checksum)
+            assert repo
+        return repo, head_commit, project
+
+    async def _upsert_snapshot(
+        self, repo: RowProxy, project: RowProxy, conn: SAConnection
+    ):
+        # has changes wrt previous commit
+        # if exists, ui might change
+        insert_stmt = pg_insert(projects_vc_snapshots).values(
+            checksum=repo.project_checksum,
+            content={"workbench": project.workbench, "ui": project.ui},
+        )
+        upsert_snapshot = insert_stmt.on_conflict_do_update(
+            constraint=projects_vc_snapshots.primary_key,
+            set_=dict(content=insert_stmt.excluded.content),
+        )
+        await conn.execute(upsert_snapshot)
+
+    # PUBLIC
+
+    async def list_repos(
+        self,
+        offset: NonNegativeInt = 0,
+        limit: Optional[PositiveInt] = None,
+    ) -> Tuple[List[RowProxy], NonNegativeInt]:
+
+        async with self.engine.acquire() as conn:
+            repo_orm = self.ReposOrm(conn)
+
+            rows: List[RowProxy]
+            rows, total_count = await repo_orm.fetch_page(
+                "project_uuid", offset=offset, limit=limit
+            )
+
+            return rows, total_count
+
+    async def get_repo_id(self, project_uuid: UUID) -> Optional[int]:
+        async with self.engine.acquire() as conn:
+            repo_orm = self.ReposOrm(conn).set_filter(project_uuid=str(project_uuid))
+            repo = await repo_orm.fetch("id")
+            return repo.id if repo else None
+
+    async def init_repo(self, project_uuid: UUID) -> int:
+        async with self.engine.acquire() as conn:
+
+            async with conn.begin():
```

Comment:
Since you are editing multiple tables, should this not be under a transaction? In case of errors you can rollback and return raise the error again.

Reply:
that is indeed with ``conn.begin()`` does, right? SEE https://aiopg.readthedocs.io/en/stable/sa.html#aiopg.sa.SAConnection.begin